feat(expensify-card): add get physical card button and routes

pac-guerreiro opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/22877
PROPOSAL: 


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Go to `Settings/Wallet`.
2. (**Optional**) 
On Web, add this mock data in browser console:

```
window.Onyx.merge(`cardList`, {
   234523452345: {
       key: '234523452345',
        cardID: 234523452345,
        state: 2,
        bank: 'Expensify Card',
        availableSpend: 10000,
        domainName: 'expensify.com',
        lastFourPAN: '2345',
        isVirtual: false,
        fraud: null,
    },
});
```
On remaining platforms, add this useEffect on WalletPage.js:

```
useEffect(() => {
        if (cardList[234523452345]) {
            return;
        }
        // eslint-disable-next-line rulesdir/prefer-actions-set-data
        Onyx.merge(`cardList`, {
            234523452345: {
                key: '234523452345',
                cardID: 234523452345,
                state: 2,
                bank: 'Expensify Card',
                availableSpend: 10000,
                domainName: 'expensify.com',
                lastFourPAN: '2345',
                isVirtual: false,
                fraud: null,
            },
        });
    }, [cardList]);
```

3. Click on the Expensify Card in the `Assigned Cards` section list.
4. On the Expensify Card page you're currently in, press the `Get Physical Card` button at the bottom of the page.
5. If you have filled your personal details before, you should be shown the confirmation page with the correct data already filled.
6. Otherwise, you'll be shown either the legal name, phone number or address steps until you have filled every single piece of data required.
7. When all data is filled, you can press the `Ship Card` button on the confirmation step to go back to the `Expensify Card` page.
8. At this point, it's not possible to send the request to the API so the `Expensify Card` page won't show the updated state when you already requested the physical card.

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

**User with no personal details**

- [ ] set up a domain with the expensify card (or use expensifail.com)
- [ ] assign a user of the domain a card
- [ ] log into the user account on newdot
- [ ] go to settings/wallet and click on the Expensify Card in the Assigned Cards  section
- [ ] verify button at the bottom says Get Physical Card
- [ ] Click on the button. Verify you’re shown  the legal name, phone number or address steps and fill them out until you have filled every single piece of data required.
- [ ] When all data is filled, you can press the Ship Card button on the confirmation step to go back to the Expensify Card page.
- [ ] Confirm that you are taken back to card page and the button reads Activate Physical Card

**User with personal details**

- [ ] set up a domain with the expensify card (or use expensifail.com)
- [ ] assign a user of the domain a card
- [ ] log into the user account on newdot
- [ ] go to settings/personal details and set the personal details for the user
- [ ] go to settings/wallet and click on the Expensify Card in the Assigned Cards  section
- [ ] verify button at the bottom says Get Physical Card
- [ ] Click on the button. Verify that you are shown the confirmation page with the correct data already filled.
- [ ] Verify you can change all three sections without error
- [ ] When all data is filled, you can press the Ship Card button on the confirmation step to go back to the Expensify Card page.
- [ ] Confirm that you are taken back to card page and the button reads Activate Physical Card

**QA Personal Details due to Refactor**

- [ ] Create a new account
- [ ] Set all the personal details under settings/profile/personal-details 
- [ ] Verify that personal details are set without error.
- [ ] Ensure that addresses in USA with the State field are saved without error as well as addresses outside of the USA.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->
Legal Name Step
<img width="406" alt="Screenshot 2023-10-23 at 16 05 01" src="https://github.com/Expensify/App/assets/48553277/45b3ae14-54b8-40c2-bcfd-1b43fab650bd">

Phone Number Step
<img width="406" alt="Screenshot 2023-10-23 at 16 05 13" src="https://github.com/Expensify/App/assets/48553277/1abee20b-1557-42b0-811a-4afe424773f6">

Address Step
<img width="406" alt="Screenshot 2023-10-23 at 16 05 22" src="https://github.com/Expensify/App/assets/48553277/05058b4d-075d-435f-afb2-61c54b5a1f7d">

Confirmation Step
<img width="406" alt="Screenshot 2023-10-23 at 16 05 30" src="https://github.com/Expensify/App/assets/48553277/35e41569-8fc2-43e0-a114-5b04deb4236c">

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

Expensify Card Page
![Screenshot_1698161864](https://github.com/Expensify/App/assets/48553277/8984fb25-d4af-4700-9b59-af402b87ed7e)

Legal Name Step
![Screenshot_1698161941](https://github.com/Expensify/App/assets/48553277/2205278c-5a41-4018-8958-cb319ec6f7a3)

Phone Number Step
![Screenshot_1698161945](https://github.com/Expensify/App/assets/48553277/310eeca6-31bf-47f6-b38a-7c38757ff0fc)

Address Step
![Screenshot_1698161949](https://github.com/Expensify/App/assets/48553277/c1ac276b-59d0-4ea5-b80d-ce1b27e4ba3e)

Confirmation Step
![Screenshot_1698161953](https://github.com/Expensify/App/assets/48553277/47e7ae8c-899b-49e8-bbb8-6486957a3e78)


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

Legal Name Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 29 51](https://github.com/Expensify/App/assets/48553277/c6de77c9-ea8e-41c6-8fae-1f4aeb61acd4)

Phone Number Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 29 56](https://github.com/Expensify/App/assets/48553277/14a45abe-12f9-4018-bea9-eaaefd66274f)

Address Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 30 03](https://github.com/Expensify/App/assets/48553277/fe199a6b-18cf-4054-8847-1830f91a2cbe)

Confirmation Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 31 49](https://github.com/Expensify/App/assets/48553277/6a72a2c2-c412-4aa5-b9d5-3f5b0609f655)

Expensify Card Page
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 28 18](https://github.com/Expensify/App/assets/48553277/71011b0f-eefc-4536-b7ef-e174a3801208)


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->
Legal Name Step
<img width="406" alt="Screenshot 2023-10-23 at 16 38 31" src="https://github.com/Expensify/App/assets/48553277/e04abe50-42b7-431e-b401-4dd3a2adf145">

Phone Number Step
<img width="406" alt="Screenshot 2023-10-23 at 16 38 47" src="https://github.com/Expensify/App/assets/48553277/6fbda728-0fed-4757-a52e-b5782ef7c15d">

Address Step
<img width="406" alt="Screenshot 2023-10-23 at 16 38 55" src="https://github.com/Expensify/App/assets/48553277/5435471f-b9d7-40e4-8eb5-0928c31748ef">

Confirmation Step
<img width="406" alt="Screenshot 2023-10-23 at 16 39 05" src="https://github.com/Expensify/App/assets/48553277/a3b04d4c-e3b4-49fc-a0f9-192e37b2bf3a">

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

Legal Name Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 21 20](https://github.com/Expensify/App/assets/48553277/ccf77efe-27b6-4fe1-aedc-76de354f7e8b)

Phone Number Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 21 25](https://github.com/Expensify/App/assets/48553277/d4fa655e-e1b5-4a4d-b948-b6206329bea6)

Address Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 21 32](https://github.com/Expensify/App/assets/48553277/cf08b53f-0ed6-4442-b40f-7988a4ab5dbf)

Confirmation Step
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 26 32](https://github.com/Expensify/App/assets/48553277/c3d7b96e-dc0b-46aa-98ff-04add3037157)

Expensify Card Page
![Simulator Screenshot - iPhone 15 Plus - 2023-10-23 at 21 21 48](https://github.com/Expensify/App/assets/48553277/18becd13-9faf-4a0e-98d1-be3257855356)

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

Expensify Card Page
![Screenshot_1698160098](https://github.com/Expensify/App/assets/48553277/578cc965-19db-4beb-bf35-50154d893e51)

Legal Name Step
![Screenshot_1698160172](https://github.com/Expensify/App/assets/48553277/30dae1bd-6140-4698-959e-0169d4c09104)

Phone Number Step
![Screenshot_1698160176](https://github.com/Expensify/App/assets/48553277/af53a16f-26ec-479f-be7f-c6082a8f34c8)

Address Step
![Screenshot_1698160182](https://github.com/Expensify/App/assets/48553277/5c70c275-c357-45e0-9acb-c6c95a9a6bb9)

Confirmation Step
![Screenshot_1698160167](https://github.com/Expensify/App/assets/48553277/a4e20f33-966f-44c0-9634-334a64e13e78)


</details>

pac-guerreiro committed:
feat(expensify-card): add get physical card button and routes

pac-guerreiro commented:
cc @grgia

pac-guerreiro commented on src/languages/en.ts:
```diff
@@ -817,6 +817,28 @@ export default {
         physicalCardNumber: 'Physical card number',
+        getPhysicalCard: 'Get physical card',
+    },
+    getPhysicalCard: {
```
@grgia 

Can I get this translated to Spanish, please?

pac-guerreiro replied:
bump @grgia

marcaaron replied:
We might have some of these already and could move them to `common`

https://github.com/Expensify/App/blob/9e35d91dd59b9233a584150a9a19f1ef458e9622/src/languages/en.ts#L1044-L1046

pac-guerreiro commented on src/libs/actions/Wallet.js:
```diff
@@ -321,6 +321,31 @@ function answerQuestionsForWallet(answers, idNumber) {
 }
 
+function requestPhysicalExpensifyCard(cardID, updatedPersonalDetails) {
+    // TODO: confirm required parameters
```
@grgia 

Can you tell me what parameters are missing here and confirm if the onyx data that is being sent it the correct one? Thanks

pac-guerreiro commented on src/pages/settings/Wallet/Cards/BaseGetPhysicalCard.js:
```diff
@@ -0,0 +1,68 @@
+        if (isConfirmation) {
+            // TODO: Use cardID and pass formatted formData here
+            Wallet.requestPhysicalExpensifyCard('', formData);
+            // TODO: Redirect user to the domain card page (?)
```
@grgia 

Should the user be redirected to the domain card page on submission?

grgia replied:
Yes, let's return to the card page.

pac-guerreiro commented on src/pages/settings/Wallet/Cards/BaseGetPhysicalCard.js:
```diff
@@ -0,0 +1,68 @@
+        >
+            <HeaderWithBackButton
+                title={title}
+                // TODO: Should form draft data be erased when a user exits the flow?
```
@grgia 

Should draft data be erased when the user exists the get physical card flow?

grgia replied:
Let's follow the same form patterns throughout the app. I believe that means no, we don't need to erase draft data

MrMuzyk commented on src/libs/Navigation/Navigation.js:
```diff
@@ -255,6 +256,39 @@ function setIsNavigationReady() {
 
+/**
+ * @param {Array<String>} loginList
+ * @returns {String} phone number, if there is one
```
if it can be undefined should we have a union here to account for that?

pasyukevich commented on src/ROUTES.ts:
```diff
@@ -95,10 +111,7 @@ export default {
-        route: 'settings/profile/contact-methods/:contactMethod/details',
-        getRoute: (contactMethod: string) => `settings/profile/contact-methods/${encodeURIComponent(contactMethod)}/details`,
-    },
+    SETTINGS_CONTACT_METHOD_DETAILS: { route: 'settings/profile/contact-methods/:contactMethod/details', getRoute: (contactMethod: string) => `settings/profile/contact-methods/${encodeURIComponent(contactMethod)}/details` },
```
redundant update

grgia reviewed (commented):
Answered questions. 

Please let me know if there's specific parts you need reviewed. It's unclear when the PR is a WIP :)

grgia commented:
assigning @sobitneupane for C+ review

pac-guerreiro commented:
> Answered questions.
> 
> Please let me know if there's specific parts you need reviewed. It's unclear when the PR is a WIP :)

Sorry @grgia, how can I make it clear for you? 😄

pac-guerreiro committed:
style: apply prettier

pac-guerreiro commented:
@grgia @joekaufmanexpensify  I require these translated to spanish:

- header: 'Get physical card',
- nameMessage: 'Enter your first and last name, as this will be shown on your card.',
- legalName: 'Legal name',
- firstName: 'Legal first name',
- lastName: 'Legal last name',
- phoneMessage: 'Enter your phone number.',
- phoneNumber: 'Phone number',
- address: 'Address',
- addressMessage: 'Enter your shipping address.',
- streetAddress: 'Street Address',
- city: 'City',
- state: 'State',
- zipPostcode: 'Zip/Postcode',
- country: 'Country',
- confirmMessage: 'Please confirm your details below.',
- estimatedDeliveryMessage: 'Your physical card will arrive in 2-3 business days.',
- next: 'Next',
- getPhysicalCard: 'Get physical card',
- shipCard: 'Ship card',`

joekaufmanexpensify commented:
@pac-guerreiro I've requested [translation assistance](https://expensify.slack.com/archives/C21FRDWCV/p1697138517668229) for these internally, and will LYK when I hear back.

joekaufmanexpensify commented:
@pac-guerreiro I was advised that a good number of these translations already exist in the app. Mind looking into that first, and let me know which ones you aren't able to find?

sobitneupane commented:
I won't be available next week. If any action is required from my end, please feel free to reassign the issue.

pac-guerreiro committed:
refactor(get physical card): rename translation keys to match onyx properties

pac-guerreiro committed:
feat(get physical card): add missing spanish translations

pac-guerreiro committed:
refactor(get physical card): reuse UI from personal details address page

pac-guerreiro committed:
fix(get physical card): wrong params passed to Navigation.goToNextPhysicalCardRoute in ExpensifyCardPage

pac-guerreiro committed:
fix(get physical card): missing testID error on console

pac-guerreiro committed:
fix(get physical card): wrong data accessed during flow

pac-guerreiro commented on src/languages/es.ts:
```diff
@@ -891,6 +892,28 @@ export default {
             thatDidntMatch: 'Los 4 últimos dígitos de tu tarjeta no coinciden. Por favor, inténtalo de nuevo.',
         },
     },
+    // TODO: add translation
```
@joekaufmanexpensify I am still missing some spanish translations that I could not find in the codebase. Could you get them for me, please? 😄

pac-guerreiro commented on src/libs/actions/Wallet.js:
```diff
@@ -321,6 +321,36 @@ function answerQuestionsForWallet(answers, idNumber) {
+        legalFirstName,
+        legalLastName,
+        phoneNumber,
+        addressCity: city,
```
@grgia I followed the specifications but I'm thinking that the country and state is missing here

marcaaron replied:
👍  agree, I think we need these. Here are the complete params the API expects

```
authToken,
legalFirstName,
legalLastName,
phoneNumber,
addressCity,
addressStreet,
addressZip,
addressCountry,
addressState
```

grgia commented:
cc @marcaaron if you have the bandwidth to take a look at this one as well

marcaaron commented on src/ROUTES.ts:
```diff
@@ -81,6 +81,22 @@ export default {
+    },
+    SETTINGS_WALLET_CARDS_GET_PHYSICAL_CONFIRM: {
+        route: '/settings/wallet/cards/:domain/get-physical/confirm',
+        getRoute: (domain: string) => `/settings/wallet/cards/${domain}/get-physical/confirm`,
```
Can we make these all `/wallet/card/` for consistency with the other routes?

marcaaron commented on src/components/AddressForm.js:
```diff
@@ -0,0 +1,192 @@
+    street1: PropTypes.string,
+    street2: PropTypes.string,
+    submitButtonText: PropTypes.string,
+    zip: PropTypes.string,
```
These please all need some comments above them like we have in other parts of the app.

grgia replied:
@pac-guerreiro could you address this comment

marcaaron commented on src/libs/Navigation/Navigation.js:
```diff
@@ -262,6 +263,31 @@ function setIsNavigationReady() {
+function goToNextPhysicalCardRoute(privatePersonalDetails, loginList) {
+    const {address, legalFirstName, legalLastName, phoneNumber} = privatePersonalDetails;
+    const currentRoute = navigationRef.current && navigationRef.current.getCurrentRoute();
+    const {domain} = (currentRoute && currentRoute.params) || {domain: ''};
```
If there can be a fallback to an empty string for `domain` should we handle it somehow? 

Seems like it would be unexpected if that happened based on the fact that we use it to navigate below.

marcaaron commented on src/libs/UserUtils.ts:
```diff
@@ -188,6 +189,14 @@ function generateAccountID(searchValue: string): number {
+ * Gets the secondary phone login number
+ */
+function getSecondaryPhoneLogin(loginList: Record<string, Login>): string | undefined {
+    const parsedLoginList = Object.keys(loginList).map((login) => login.replace('@expensify.sms', ''));
```
We can use the amazing `Str.removeSMSDomain()` here.

marcaaron commented on src/libs/actions/Wallet.js:
```diff
@@ -321,6 +321,36 @@ function answerQuestionsForWallet(answers, idNumber) {
+            {
+                onyxMethod: Onyx.METHOD.MERGE,
+                key: ONYXKEYS.PRIVATE_PERSONAL_DETAILS,
+                value: params,
```
I am not sure if these params can be dropped here as is...

Especially the `address` you can see here it has this format:

https://github.com/Expensify/App/blob/9e35d91dd59b9233a584150a9a19f1ef458e9622/src/types/onyx/PrivatePersonalDetails.ts#L1-L16

So what we are storing might be different than what we send to the API.

pac-guerreiro replied:
Besides that, you're right, I need to send the data directly without mapping it

pac-guerreiro replied:
@shawnborton The designs for the confirmation page don't have a fixed size applied to the items of the list, as you can see below:
![Screenshot 2023-10-18 at 12 02 32](https://github.com/Expensify/App/assets/48553277/089e4d46-761c-4dbd-8397-30623f93c03c)

Should these items size match the existing ones on the app? And if so, could you update the designs, please?

shawnborton replied:
@pac-guerreiro those should just be existing components in the design system called "push-to-page" rows. They might be called something like MenuOption in the code? cc @grgia for help there

shawnborton replied:
But you should be able to reuse existing code for these and not need to worry about building something new.

pac-guerreiro replied:
@shawnborton thanks for the response! I'm already reusing the right component for this but I was applying the styles from Figma, so I removed these styles and tweaked the whole layout to work correctly 😄 

This is the current state:

![Screenshot 2023-10-18 at 12 56 26](https://github.com/Expensify/App/assets/48553277/14563216-eee4-4a8c-8841-a9369ce569a3)

shawnborton replied:
Sweet, that feels pretty good to me!

marcaaron reviewed (commented):
Great refactoring and nice work so far. Just leaving some initial comments to get us started. Thanks for the work on this!

pac-guerreiro committed:
fix(wallet): wrong naming for domain card route

pac-guerreiro committed:
fix(get physical card): top margin on address page step

pac-guerreiro committed:
fix(get physical card): wrong prop type supplied to children

pac-guerreiro committed:
refactor(get physical card): rename routes

pac-guerreiro committed:
fix(get physical card): wrong data sent through final API request

pac-guerreiro committed:
refactor(get physical card): replace manual solution with existing util

pac-guerreiro committed:
fix(get physical card): wrong layout on confirm page step

shawnborton commented:
Feels pretty good to me!

shawnborton commented:
> Should there be some visual feedback to the user like an alert saying that the physical card was requested?

In your video, what happens after you press Ship Card?

shawnborton commented:
Can you show what that looks like? Because that might help us answer your question. Basically once that happens, does the wallet page change in any way to give feedback that a card was ordered?

pac-guerreiro commented:
> Can you show what that looks like? Because that might help us answer your question. Basically once that happens, does the wallet page change in any way to give feedback that a card was ordered?

@shawnborton I don't have the means to test the API right now, so I'm not sure what changes happen to the internal state of the app after the API call 😅

shawnborton commented:
Ah okay. Let's hold off until we move onto that phase then.

grgia commented:
@pac-guerreiro 
> Should there be some visual feedback to the user like an alert saying that the physical card was requested?


The visual feedback right now is that after completing the flow, the button will say `Activate Card`

![image](https://github.com/Expensify/App/assets/38015950/7f50d4fa-35be-4292-8920-2083aa4f04c6)

pac-guerreiro commented:
> @pac-guerreiro
> 
> > Should there be some visual feedback to the user like an alert saying that the physical card was requested?
> 
> The visual feedback right now is that after completing the flow, the button will say `Activate Card`
> 
> ![image](https://user-images.githubusercontent.com/38015950/276383824-7f50d4fa-35be-4292-8920-2083aa4f04c6.png)

@grgia thank you! Regarding the second question I left above, what do you think ?

grgia commented:
@pac-guerreiro could you resend the question?

grgia commented:
@pac-guerreiro I've assigned @allroundexperts for C+ review. Would you kindly let us know if this PR is ready for review?

grgia commented:
We need to include QA steps for testing the personal details refactor as well. Please make sure you include test cases to cover all changes when you are ready. @pac-guerreiro

pac-guerreiro committed:
feat(get physical card): complete Spanish translation

pac-guerreiro commented:
This PR is still a WIP, so I'll let you know when it's ready to be reviewed! cc @grgia @joekaufmanexpensify @allroundexperts

pac-guerreiro committed:
refactor: code style

pac-guerreiro committed:
feat(get physical card): use draft data for temporary modifications and then use it to submit physical card request

pac-guerreiro committed:
refactor(get physical card): phone number validation logic

pac-guerreiro committed:
fix(get physical card): wrong address displayed on confirmation page

pac-guerreiro committed:
chore(get physical card): add expensify card mock for testing purposes

grgia commented:
@allroundexperts Could you possibly prioritize this PR this week? Looks like it's ready for a round of review.

pac-guerreiro committed:
fix(get physical card): formatted address showing undefined

pac-guerreiro committed:
fix(get physical card): wrong condition to render get physical card button

grgia commented:
@allroundexperts to my knowledge yes, I was updated in slack

pac-guerreiro committed:
fix(get physical card): wrong mock data

pac-guerreiro commented:
@allroundexperts @grgia Just removed the WIP from the title! Have fun!

pac-guerreiro commented:
Just fixed an error that was thrown on the wallet page because of the list of assigned cards being nested inside a ScrollView

pac-guerreiro committed:
fix(get physical card): wrong validation of address while getting next route

pac-guerreiro committed:
fix(get physical card): validate callback not returning an object causes an error

pac-guerreiro commented:
Our QA, Marta, found some issues while testing the PR this morning:

- If address was missing, the navigation logic would navigate to the confirm page nonetheless.
- Validation callback passed to the Form component was not returning an object, so pressing Ship Card would not do anything and an error would be shown on the console.

I just fixed these errors and pushed the fixes to the branch 😄

mart4g commented:
**TESTS SUMMARY**
This feature is verified on Google Chrome Version 118.0.5993.88 (Official Build) (arm64), Safari Version Version 17.0 (19616.1.27.211.1), Xcode 15.0 -Simulator: iPhone SE (3rd generation), Version 15.0 (1015.2) + Safari iOS, Android real device: Samsung Galaxy A32  SM-A325F/DS

This feature works as expected. PR is ready to be merged.
The order of the screens depends on the data. If the address is already saved by the app, then the Enter your shipping address screen. won't display, unless a user would like to update the address form. Then this screen is displayed as the one before last.
If the address is being provided for the first time then the flow matches Figma flow.
Please see the recordings and screenshots attached. 


https://github.com/Expensify/App/assets/147029649/90574b84-40ab-4543-8fc6-37239baccf1b

https://github.com/Expensify/App/assets/147029649/c10009b6-d037-4bcc-903a-8d5bdfd66ff8

![adres already in the system flow](https://github.com/Expensify/App/assets/147029649/e59adebf-b075-4709-86e9-ca32beb077a0)
![no adress provided yet flow](https://github.com/Expensify/App/assets/147029649/70bcfae8-e315-436a-9e68-2f223f47f68e)

pac-guerreiro committed:
fix(TextInput): container styles not applied to text input help message

pac-guerreiro committed:
fix(get physical card): broken phone number validation

pac-guerreiro commented:
@allroundexperts fixes pushed!

joekaufmanexpensify commented:
@allroundexperts as an fyi this one is now ready for review again!

pac-guerreiro committed:
chore(get physical card): apply prettier

grgia commented:
@pac-guerreiro could you also add QA steps to cover the changes to the personal details flow?

@allroundexperts would you mind taking a careful look at that flow as well with your review?

pac-guerreiro committed:
refactor(get physical card): separate utils from core navigation

pac-guerreiro committed:
fix(get physical card): RHP getting closed on back button pressed

pac-guerreiro committed:
fix(get physical card): being able to access a step before completing previous one

pac-guerreiro committed:
fix(get physical card): navigating from confirmation page not pushing the screen into stack

pac-guerreiro committed:
fix(get physical card): user getting redirected to confirm page when trying to go to a step

pac-guerreiro committed:
fix(get physical card): redirect not replacing current screen on navigation stack

pac-guerreiro committed:
fix(get physical card): draft values initialization

pac-guerreiro commented:
@allroundexperts @joekaufmanexpensify @grgia Branch rebased, conflicts resolved and all fixes addressed. PR is ready for review!

pac-guerreiro committed:
fix: invalid types on setDraftValues

pac-guerreiro committed:
refactor: migrate FormActions to Typescript

pac-guerreiro commented:
> @pac-guerreiro Type checks seem to be failing?

@allroundexperts 

Should pass now!

pac-guerreiro commented:
@allroundexperts 

I refactored the address form on `Settings > Profile > Personal Details > Address`, so that it could be reused on this flow. Could you give it a spin to see if is there a problem present?

pac-guerreiro committed:
fix: add missing getPhysicalCardFormDraft

pac-guerreiro committed:
fix(get physical card): add missing onyx value types

dylanexpensify commented:
bump @allroundexperts

dylanexpensify commented:
@allroundexperts can we do this one now? This is blocking a critical issue

pac-guerreiro commented:
@dylanexpensify @grgia conflicts resolved

grgia commented on src/pages/settings/Wallet/WalletPage/WalletPage.js:
```diff
@@ -257,6 +257,27 @@ function WalletPage({bankAccountList, betas, cardList, fundList, isLoadingPaymen
         Navigation.navigate(source === CONST.KYC_WALL_SOURCE.ENABLE_WALLET ? ROUTES.SETTINGS_WALLET : ROUTES.SETTINGS_WALLET_TRANSFER_BALANCE);
     };
 
+    // TODO: REMOVE THIS AFTER TESTING
+    useEffect(() => {
+        if (cardList[234523452345]) {
+            return;
+        }
+        // eslint-disable-next-line rulesdir/prefer-actions-set-data
+        Onyx.merge(`cardList`, {
+            234523452345: {
+                key: '234523452345',
+                cardID: 234523452345,
+                state: 2,
+                bank: 'Expensify Card',
+                availableSpend: 10000,
+                domainName: 'expensify.com',
+                lastFourPAN: '2345',
+                isVirtual: false,
+                fraud: null,
+            },
+        });
+    }, [cardList]);
+
```
Let's remove this if you are ready for final review

pac-guerreiro committed:
chore(get physical card): removed mock data

pac-guerreiro committed:
doc(get physical card): add better comments to address fields

pac-guerreiro commented:
@grgia just applied the changes that you requested 😄 

@allroundexperts I updated the PR description with instructions for adding mock data if you need it!

PR is ready for final review!

pac-guerreiro committed:
chore(get physical card): remove unused import

dylanexpensify commented:
@allroundexperts mind prioritizing this first thing? 🙇‍♂️

pac-guerreiro commented:
@dylanexpensify @grgia @allroundexperts conflicts addressed.

dylanexpensify commented:
@allroundexperts @grgia can we get this re-reviewed/merged today?? 🤞

pac-guerreiro committed:
Merge branch 'main' into feature/add-get-physical-card-button-and-necessary-routes

pac-guerreiro commented:
@allroundexperts @grgia 

Here is proof of the fixes I applied to the comments you mentioned above:

- Draft data is persisted when returning to the flow

https://github.com/Expensify/App/assets/48553277/eed9b6f2-bd2b-48fa-9de1-c8ca07072d02

- Unable to access unauthorized steps through URL and pressing go back gets user back to confirmation page

https://github.com/Expensify/App/assets/48553277/4d0fbc16-0f10-409a-bfc6-19fa861ef081

pac-guerreiro committed:
chore: remove unused file

pac-guerreiro commented:
I just removed a `.js` file that got migrated to `.ts` extension in `main` branch, since it was a duplicate and was causing typescript checks fails

pac-guerreiro committed:
refactor: use form utils to get draft key

grgia reviewed (commented):
Can you update your description to move what you have under QA to the tests section and add these steps to QA

```
**User with no personal details**

- [ ] set up a domain with the expensify card (or use expensifail.com)
- [ ] assign a user of the domain a card
- [ ] log into the user account on newdot
- [ ] go to settings/wallet and click on the Expensify Card in the Assigned Cards  section
- [ ] verify button at the bottom says Get Physical Card
- [ ] Click on the button. Verify you’re shown  the legal name, phone number or address steps and fill them out until you have filled every single piece of data required.
- [ ] When all data is filled, you can press the Ship Card button on the confirmation step to go back to the Expensify Card page.
- [ ] Confirm that you are taken back to card page and the button reads Activate Physical Card

**User with personal details**

- [ ] set up a domain with the expensify card (or use expensifail.com)
- [ ] assign a user of the domain a card
- [ ] log into the user account on newdot
- [ ] go to settings/personal details and set the personal details for the user
- [ ] go to settings/wallet and click on the Expensify Card in the Assigned Cards  section
- [ ] verify button at the bottom says Get Physical Card
- [ ] Click on the button. Verify that you are shown the confirmation page with the correct data already filled.
- [ ] Verify you can change all three sections without error
- [ ] When all data is filled, you can press the Ship Card button on the confirmation step to go back to the Expensify Card page.
- [ ] Confirm that you are taken back to card page and the button reads Activate Physical Card

**QA Personal Details due to Refactor**

- [ ] Create a new account
- [ ] Set all the personal details under settings/profile/personal-details 
- [ ] Verify that personal details are set without error.
- [ ] Ensure that addresses in USA with the State field are saved without error as well as addresses outside of the USA.
```

Please add more steps that cover all your changes for our QA team

pac-guerreiro commented:
@grgia description updated!

pac-guerreiro committed:
feat(get physical card): enable physical card request to API

pac-guerreiro commented on src/libs/actions/Wallet.js:
```diff
@@ -330,6 +330,45 @@ function answerQuestionsForWallet(answers, idNumber) {
+            },
+        ],
+    };
+    API.write('RequestPhysicalExpensifyCard', params, onyxData);
```
@grgia This is the physical card request API call

pac-guerreiro committed:
Merge branch 'main' into feature/add-get-physical-card-button-and-necessary-routes

# Conflicts:
#	src/pages/settings/Profile/PersonalDetails/AddressPage.js

pac-guerreiro commented:
@grgia @allroundexperts @dylanexpensify conflicts resolved

pac-guerreiro commented:
> **Another Bug**
> 
> 
> 
> If you try to open the confirm page directly via URL and try to go back, the `Not found` page is shown.
> 
> 
> 
> https://github.com/Expensify/App/assets/30054992/054e83f2-c902-404c-80ef-8a6296176bee

@allroundexperts were you using the mock data that I shared on the PR description? If so, than that's perfectly normal because the data gets erased when you go back because the data is refreshed with what's on the backend. You can comprove this by doing this again and setting the mock data on the console on the not found page.

You did encounter a weird behaviour though that I did not account for which is accessing the flow through URL while outside of the flow. You were able to access the confirmation page although the draft data was not initialized. I'm working on a fix for this behaviour and I should be able to push it today.

cc @grgia @dylanexpensify

pac-guerreiro committed:
fix(get physical card): being able to access the flow with the draft data not initialized

pac-guerreiro committed:
fix(get physical card): being able to access the flow when the domain has no cards

pac-guerreiro committed:
fix(get physical card): ability to access the flow when there is no physical card or it doesnt have the necessary state

pac-guerreiro committed:
fix(expensify card): pressing go back button on not found state closes the RHP

pac-guerreiro committed:
fix(get physical card): wrong accessibility prop names

pac-guerreiro committed:
Merge branch 'main' into feature/add-get-physical-card-button-and-necessary-routes

pac-guerreiro commented:
@grgia @allroundexperts @dylanexpensify 

All reported issues should be fixed now!

I added logic to handle these edge cases:
- When the specified domain has no cards, the user is redirected to the wallet page;
- When the physical card does not exist on the specified domain, or if does exist but it doesn't have the required state for the flow, then the user is redirected to the card page of the specified domain;

Also, I noticed that while on the not found state of the card page if the user presses the back button then the RHP is closed. So I changed this behaviour to redirect the user to the wallet page instead.

pac-guerreiro committed:
style: apply linter

grgia commented:
@allroundexperts bump

grgia commented:
@allroundexperts I agree, if no bugs are blocking we can handle in a follow up. As you finish checklist and QA steps, can you please catalogue the non-blocking bugs so I can create issues

pac-guerreiro commented on src/components/Form.js:
```diff
@@ -303,7 +304,8 @@ function Form(props) {
                 // We want to initialize the input value if it's undefined
                 if (_.isUndefined(inputValues[inputID])) {
-                    inputValues[inputID] = _.isBoolean(defaultValue) ? defaultValue : defaultValue || '';
+                    // eslint-disable-next-line es/no-nullish-coalescing-operators
+                    inputValues[inputID] = defaultValue ?? '';
```
I made this change because it's more simpler and does the same thing as before.

grgia replied:
<img width="424" alt="image" src="https://github.com/Expensify/App/assets/38015950/f34dafd9-17d2-48eb-8c34-6ae93cb13412">

@pac-guerreiro @allroundexperts Shall we revert this line to avoid breaking other forms?

grgia replied:
<img width="436" alt="image" src="https://github.com/Expensify/App/assets/38015950/77a6faf9-34e0-4dc9-9d6b-e509bff5775f">

although @allroundexperts is this the case you were referring to?

pac-guerreiro commented:
@allroundexperts question answered

allroundexperts reviewed (approved):
Looks good!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6951287204) to staging by https://github.com/grgia in version: 1.4.2-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

situchan commented on src/libs/Navigation/linkTo.js:
```diff
@@ -83,6 +83,17 @@ export default function linkTo(navigation, path, type) {
     if (action.payload.name === NAVIGATORS.RIGHT_MODAL_NAVIGATOR) {
         const minimalAction = getMinimalAction(action, navigation.getRootState());
         if (minimalAction) {
+            // There are situations where a route already exists on the current navigation stack
+            // But we want to push the same route instead of going back in the stack
+            // Which would break the user navigation history
+            if (type === CONST.NAVIGATION.ACTION_TYPE.PUSH) {
+                minimalAction.type = CONST.NAVIGATION.ACTION_TYPE.PUSH;
+            }
+            // There are situations when the user is trying to access a route which he has no access to
+            // So we want to redirect him to the right one and replace the one he tried to access
+            if (type === CONST.NAVIGATION.ACTION_TYPE.REPLACE) {
+                minimalAction.type = CONST.NAVIGATION.ACTION_TYPE.REPLACE;
+            }
```
This caused regression - https://github.com/Expensify/App/issues/31688
What issue this code part is trying to fix? Is it more critical than https://github.com/Expensify/App/issues/31688?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6972415425) to production by https://github.com/luacmartins in version: 1.4.2-3 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

Pujan92 commented on src/components/TextInput/BaseTextInput/index.js:
```diff
@@ -250,7 +250,7 @@ function BaseTextInput(props) {
         <>
             <View
-                style={styles.pointerEventsNone}
+                style={[styles.pointerEventsNone, ...props.containerStyles]}
```
Moved `containerStyles` to main `View` wrapper caused an issue https://github.com/Expensify/App/issues/40519, bcoz `containerStyles` has max height restriction which specifically required only for TextInput Due to this change the error message overflows and overlaps with bottom content.

DylanDylann commented on src/pages/settings/Wallet/Card/BaseGetPhysicalCard.js:
```diff
@@ -0,0 +1,236 @@
+    return (
+        <ScreenWrapper
+            shouldEnablePickerAvoiding={false}
+            shouldShowOfflineIndicator={false}
```
It caused [this bug](https://github.com/Expensify/App/issues/54467), we need to remove shouldShowOfflineIndicator props
